src/jsTyping/jsTyping.ts

```diff
@@ -214,17 +204,99 @@ namespace ts.JsTyping {
+            }
+
+            // Now we scan the directories for typing information in
+            // already-installed dependencies (if present). Note that this
+            // step happens regardless of whether a manifest was present,
+            // which is certainly a valid configuration, if an unusual one.
+            const packagesFolderPath = combinePaths(projectRootPath, modulesDirName);
+            filesToWatch.push(packagesFolderPath);
+            if (!host.directoryExists(packagesFolderPath)) {
                 return;
             }
 
-            filesToWatch.push(jsonPath);
-            const jsonConfig: PackageJson = readConfigFile(jsonPath, path => host.readFile(path)).config;
-            const jsonTypingNames = flatMap([jsonConfig.dependencies, jsonConfig.devDependencies, jsonConfig.optionalDependencies, jsonConfig.peerDependencies], getOwnKeys);
-            addInferredTypings(jsonTypingNames, `Typing names in '${jsonPath}' dependencies`);
+            // There's two cases we have to take into account here:
+            // 1. If manifest is undefined, then we're not using a manifest.
+            //    That means that we should scan _all_ dependencies at the top
+            //    level of the modulesDir.
+            // 2. If manifest is defined, then we can do some special
+            //    filtering to reduce the amount of scanning we need to do.
+            //
+            // Previous versions of this algorithm checked for a `_requiredBy`
+            // field in the package.json, but that field is only present in
+            // `npm@>=3 <7`.
+
+            // Package names that do **not** provide their own typings, so
+            // we'll look them up.
+            const packageNames: string[] = [];
+
+            const dependencyManifestNames = manifestTypingNames
+                // This is #1 described above.
+                ? manifestTypingNames.map(typingName => combinePaths(packagesFolderPath, typingName, manifestName))
+                // And #2. Depth = 3 because scoped packages look like `node_modules/@foo/bar/package.json`
+                : host.readDirectory(packagesFolderPath, [Extension.Json], /*excludes*/ undefined, /*includes*/ undefined, /*depth*/ 3)
+                    .filter(manifestPath => {
+                        if (getBaseFileName(manifestPath) !== manifestName) {
+                            return false;
+                        }
+                        // It's ok to treat
+                        // `node_modules/@foo/bar/package.json` as a manifest,
+                        // but not `node_modules/jquery/nested/package.json`.
+                        // We only assume depth 3 is ok for formally scoped
+                        // packages. So that needs this dance here.
+                        const pathComponents = getPathComponents(normalizePath(manifestPath));
+                        const isScoped = pathComponents[pathComponents.length - 3][0] === "@";

```

Comment:
Consider adding a length check for `pathComponents` so we don't throw on a path such as `/node_modules/package.json`. (No good reason that should exist, but we shouldn't crash on wacky input either)

Reply:
Oh! Didn't realize `/` would be its own component. Thanks for setting me straight

Reply:
I was also very surprised 😂